feat(metrics): Extract transaction metrics in external relays [INGEST-1477]

iker-barriocanal opened the pull request:
Currently, transaction metrics are only extracted in processing relays (the function call to extract metrics is surrounded by an `if_processing!`). Even if we extract metrics in external relays, we don't want to extract metrics multiple times from a transaction, causing metric (and billing) duplication.

The implemented approach in this PR is to extract metrics in the very first relay in the chain, mark the transaction as extracted, and not extract metrics from that transaction in subsequent relays. The simplest way to accomplish that is using the existing `metrics_extracted` header in items that had metrics extracted. The approach and the header are reused from the metric extraction from sessions, there's nothing new here.

Implementation-wise, removing that `if_processing` triggers a lot of changes to make current processing-only features non-processing too. On the other hand, we store a new flag `transaction_metrics_extracted` on `ProcessEnvelopeState` to make it available to the whole state processing function and transfer it to the outgoing event. Ideally, `process_state` is refactored not to throw every piece of data we need to the envelope state object, but that's out of the scope of this PR.

iker-barriocanal committed:
Extract non-processing imports in envelopes actor

Move all non-processing imports out of the processing cfg config in the
envelopes actor.

iker-barriocanal committed:
Don't extract transactions when extracted header is present

The `metrics_extracted` header must only be used on items that had
transactions extracted. This means that relay must not extract metrics
from items with such header, to not duplicate metrics.

iker-barriocanal committed:
Get config from state using the function

Use the function to get the config, instead of accessing the struct's
property directly.

iker-barriocanal committed:
Check if tx metrics config matches required version

iker-barriocanal committed:
Ensure relay must extract metrics before extracting config

iker-barriocanal committed:
Mark item as metrics_extracted right after extracting items

iker-barriocanal committed:
Use existing tx item in envelope serialization

iker-barriocanal committed:
Add an integration test

iker-barriocanal committed:
Add version in tx metrics config

iker-barriocanal committed:
Fix: set payload for all items, including txs\

The transaction item was reused to keep the item's headers. However, no
additional payload was being set and this was causing missing data such
as missing breakdowns.

iker-barriocanal committed:
test: ensure no extra events were sent

iker-barriocanal committed:
test: ensure metrics_extracted header exists

iker-barriocanal committed:
Split big integration test into two smaller tests

iker-barriocanal committed:
Merge branch 'master' into iker/feat/mep-support-ext-relays

iker-barriocanal committed:
Add test: use too low and big versions

iker-barriocanal committed:
Update changelog

untitaker commented on relay-server/src/actors/envelopes.rs:
```diff
@@ -1681,41 +1686,53 @@ impl EnvelopeProcessor {
+                .metric_conditional_tagging
+                .as_slice();
+
+            let (event, _) = self.event_from_json_payload(item, Some(EventType::Transaction))?;
```
I don't think this should be necessary. Don't you have access to `state.event` here?

jjbayer replied:
Why parse the event again here?

untitaker commented on relay-server/src/actors/envelopes.rs:
```diff
@@ -346,6 +346,8 @@ struct ProcessEnvelopeState {
     /// extracted.
     event: Annotated<Event>,
 
+    transaction_item: Option<Item>,
```
you can probably store item headers only here, not sure if it's feasible though

jjbayer commented on relay-server/src/actors/envelopes.rs:
```diff
@@ -1805,8 +1822,10 @@ impl EnvelopeProcessor {
-        let event_type = state.event_type().unwrap_or_default();
-        let mut event_item = Item::new(ItemType::from_event_type(event_type));
+        let mut event_item = match state.event_type().unwrap_or_default() {
+            EventType::Transaction => state.transaction_item.take().unwrap(),
```
There's usually a better way than `unwrap()`. E.g. store `event_item` on the state instead of `event_type` and `transaction_item`, and then get the type from the item.

jjbayer commented on relay-server/src/metrics_extraction/transactions.rs:
```diff
@@ -51,20 +49,31 @@ pub struct CustomMeasurementConfig {
-#[cfg(feature = "processing")]
+impl TransactionMetricsConfig {
+    pub fn is_enabled(&self) -> bool {
+        self.version > 0 && self.version <= EXTRACT_MAX_VERSION
```
Quick thought: We could actually remove the `self.version > 0` check and set `EXTRACT_MAX_VERSION = 0`. That way, we would not need any changes on the sentry side. Or am I missing something crucial?

jjbayer replied:
Actually, let's keep it as-is to be consistent with [session metrics extraction](https://github.com/getsentry/relay/blob/0576be7d503f8155e58cda1492216ab64631b05e/relay-server/src/metrics_extraction/sessions.rs#L35).

jjbayer commented on relay-server/src/actors/envelopes.rs:
```diff
@@ -1805,8 +1822,10 @@ impl EnvelopeProcessor {
+        let mut event_item = match state.event_type().unwrap_or_default() {
+            EventType::Transaction => state.transaction_item.take().unwrap(),
+            ty => Item::new(ItemType::from_event_type(ty)),
+        };
```
I assume we reuse the item here because we do not want to lose the `metrics_extracted` information. But this could have side effects (I don't know what else is in that item header). So it might be safer to explicitly transfer the `metrics_extracted` information to the new item, as we do with `sample_rates` a few lines below.

jjbayer committed:
ref(metrics): Simplify metrics_extracted logic

Two minor tweaks to #1344, based on comments on that PR:

- Do not parse the event payload twice.
- Store a boolean on the envelope state, rather than the full event
  item.

jjbayer committed:
ref: Minimize diff

jjbayer committed:
Merge remote-tracking branch 'origin/master' into iker/feat/mep-support-ext-relays
